### PR TITLE
Drive base pid

### PR DESCRIPTION
**Pull request by Wallace-Watler**
Completes #16 Story Issue Number

Tests Run On:
 - [x] Test Bench
 - [ ] Robot Prototype
 - [ ] Final Robot

## For Reviewers

Here are some things to look for:
 * Comments in the code.
   - Comments explain *why* the code is doing something, not what or how.
* No "magic numbers" or hard coded constants.
* Small code change (i.e. 10-15 lines or so)
* The code implements the design specified in the corresponding story.
* The code interfaces with other code and hardware properly.
  - Ex. Code implementing a command for one subsystem should not affect another subsystem.
  - Ex. Objects representing hardware on the robot should only be accessed in a subsystem.
  - Ex. Joysticks should only be accessed in setting up buttons or in commands (not subsystems).

@Noah-Kennedy
@NicoletFEAR/mentor

### CODE REVIEW

File: Steamworks/src/org/usfirst/frc/team4786/robot/Robot.java [hunk 1]
```diff
@@ -32,9 +33,8 @@
 	@Override
 	public void robotInit() {
 		oi = new OI();
-//		chooser.addDefault("Default Auto", new ExampleCommand());
+		chooser.addObject("Test PID", new DriveToPosition(1));
```
Comment (Noah-Kennedy) [inline Steamworks/src/org/usfirst/frc/team4786/robot/Robot.java]: This is fine now, but we will need to set it to display something other than "Test PID" later.

Comment (Noah-Kennedy) [inline Steamworks/src/org/usfirst/frc/team4786/robot/Robot.java]: Scratch my earlier comment. The entire sendableChooser mechanism is redundant, as it is not used anywhere else. The chooser was bypassed for running the autonomous command. Again, I think that this is fine for now, but I will write it into an issue if this passes review.

File: Steamworks/src/org/usfirst/frc/team4786/robot/RobotMap.java [hunk 2]
```diff
@@ -10,18 +10,47 @@
...
+	public static final double LeftP = 0.355 / 7; //35% throttle within 5 rotations of target
+	public static final double LeftI = RobotMap.LeftP / 100000;
+	public static final double LeftD = 0.0;
+	public static final double LeftF = 0.0;
+	//Right GearBox PID Constants
+	public static final double RightP = 0.355 / 7;
+	public static final double RightI = RobotMap.RightP / 100000;
+	public static final double RightD = 0.0;
+	public static final double RightF = 0.0;
+
```
Comment (Noah-Kennedy) [inline Steamworks/src/org/usfirst/frc/team4786/robot/RobotMap.java]: Comments look good.

File: Steamworks/src/org/usfirst/frc/team4786/robot/RobotMap.java [hunk 3]
```diff
@@ -10,18 +10,47 @@
...
+	 * Find the encoder codes per rev
+	 * If the encoder is a quad encoder, the formula for determining the starting P value is the following:
+	 * 
+	 * P = percentThrottleAsDecimal * 1023 / (4 * revolutionsBeforeSetpoint * encoderCodesPerRev)
+	 * 
+	 * For example, the P value for 25% throttle, 5 revolutions before the setpoint, with 400 codes per rev,
+	 * the equation looks like this:
+	 * 
+	 * P = .25 * 1023 / (4 * 5 * 400) = 0.03196875
+	 */
```
Comment (Noah-Kennedy) [inline Steamworks/src/org/usfirst/frc/team4786/robot/RobotMap.java]: Again, nice comments.

File: Steamworks/src/org/usfirst/frc/team4786/robot/subsystems/DriveTrain.java [hunk 4]
```diff
@@ -1,32 +1,60 @@
...
-	CANTalon frontLeft = new CANTalon(RobotMap.frontLeftPort);
-	CANTalon frontRight = new CANTalon(RobotMap.frontRightPort);
+	//CANTalon objects, there is a reason why they are private
+	private CANTalon backLeft = new CANTalon(RobotMap.backLeftPort);
+	private CANTalon backRight = new CANTalon(RobotMap.backRightPort);
+	private CANTalon frontLeft = new CANTalon(RobotMap.frontLeftPort);
+	private CANTalon frontRight = new CANTalon(RobotMap.frontRightPort);
+	
+	//Set Important PID Variables
+	//public int error = (int) (rawCodesPerRev * gearBoxRatio / RobotMap.ERROR_CONSTANT);
```
Comment (Noah-Kennedy) [inline Steamworks/src/org/usfirst/frc/team4786/robot/subsystems/DriveTrain.java]: Excellent. Code permissions change was needed.

File: Steamworks/src/org/usfirst/frc/team4786/robot/subsystems/DriveTrain.java [hunk 5]
```diff
@@ -1,32 +1,60 @@
...
 		//this inverts the cantalons on the right side
+		frontLeft.setInverted(true);
+		
+		//Beginning of the world of PID!
+		frontLeft.setProfile(0);
+		frontRight.setProfile(0);
+		frontLeft.setPID(RobotMap.LeftP, RobotMap.LeftI, RobotMap.LeftD, RobotMap.LeftF, 0, 0, 0);		
+		frontRight.setPID(RobotMap.RightP, RobotMap.RightI, RobotMap.RightD, RobotMap.RightF, 0, 0, 0);
+		frontLeft.setCloseLoopRampRate(0.125 / 8);
+		frontRight.setCloseLoopRampRate(0.125 / 8);
```
Comment (Noah-Kennedy) [inline Steamworks/src/org/usfirst/frc/team4786/robot/subsystems/DriveTrain.java]: These hardcoded numbers need to be moved to the RobotMap class. This needs to be done before the pull request is approved.

File: Steamworks/src/org/usfirst/frc/team4786/robot/subsystems/DriveTrain.java [hunk 6]
```diff
@@ -43,9 +71,58 @@ public void brake(){
...
+	//Begin PID Functions
+	
+	public void driveToPosition(double distanceToDrive){
+		//Run convertToRotations function
+		double rot = convertToRotations(distanceToDrive);
+		
+		//Change Talon modes to "position" just in case
+    	//they were in another mode before
+    	frontLeft.changeControlMode(TalonControlMode.Position);
+    	frontRight.changeControlMode(TalonControlMode.Position);
```
Comment (Noah-Kennedy) [inline Steamworks/src/org/usfirst/frc/team4786/robot/subsystems/DriveTrain.java]: Fix this whitespace.

File: Steamworks/src/org/usfirst/frc/team4786/robot/Robot.java (hunk 1 above)
Comment (jkoritzinsky) [inline Steamworks/src/org/usfirst/frc/team4786/robot/Robot.java]: I agree with Noah. The `SendableChooser` should be removed until we actually need to use it.

File: Steamworks/src/org/usfirst/frc/team4786/robot/RobotMap.java [hunk 7]
```diff
@@ -10,18 +10,47 @@
...
 	//Scaling used only for open loop drive system
 	public static final double openLoopSpeedScaling = .6;
 	
+	//General PID Constants
+	//public static final double ERROR_CONSTANT = 45;
+	public static final int DRIVETRAIN_ENCODER_CODES_PER_REV = 360;
+	//Wheel Radius measured in feet
+	public static final double WHEEL_RADIUS = 0.25;
+	//Left GearBox PID Constants
+	public static final double LeftP = 0.355 / 7; //35% throttle within 5 rotations of target
```
Comment (jkoritzinsky) [inline Steamworks/src/org/usfirst/frc/team4786/robot/RobotMap.java]: This doesn't need to be fixed now, but when the speed/velocity PID constants are added, these should be renamed to include that they have position information in them (like `LeftPositionP` or something).

File: Steamworks/src/org/usfirst/frc/team4786/robot/commands/DriveToPosition.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    // Called just before this Command runs the first time
+    protected void initialize() {
+    }
+
+    // Called repeatedly when this Command is scheduled to run
+    protected void execute() {
+    	//Let's drive!
+    	Robot.driveTrain.driveToPosition(targetPosition);
```
Comment (jkoritzinsky) [inline Steamworks/src/org/usfirst/frc/team4786/robot/commands/DriveToPosition.java]: One way to make this work a little better (and to be able to use this after moving without this command, you can move this function call to the `initialize()` method).

File: Steamworks/src/org/usfirst/frc/team4786/robot/commands/DriveToPosition.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    // Make this return true when this Command no longer needs to run execute()
+    protected boolean isFinished() {
+        return false;
```
Comment (jkoritzinsky) [inline Steamworks/src/org/usfirst/frc/team4786/robot/commands/DriveToPosition.java]: Eventually (when you make the gear placement auto program) you will have to make this return true when the drive train has traveled to the position. Once again this doesn't have to be done now.

File: Steamworks/src/org/usfirst/frc/team4786/robot/subsystems/DriveTrain.java [hunk 10]
```diff
@@ -43,9 +71,58 @@ public void brake(){
...
+		//Make sure we inverse this right side,
+		//otherwise, you have a spinning robot on your hands
+		frontRight.set(-rot);
+		
+		SmartDashboard.putNumber("Position", frontLeft.getPosition());
+	}
+	
+	//Have these functions to get CANTalon objects
+	//since they are private
+	public CANTalon getFrontLeft(){
```
Comment (jkoritzinsky) [inline Steamworks/src/org/usfirst/frc/team4786/robot/subsystems/DriveTrain.java]: You should remove these accessor functions. You should try to keep the talon srx motors as purely an implementation detail of the subsystem and not exposed to other code.

**Comment by jkoritzinsky** [comment]
looks good to me!

**Pull request merged**